File: lessons/dsa/two-pointers/examples/find_pair_with_sum_test.go
package examples

import "testing"

func TestFindPairWithSum(t *testing.T) {
	cases := []struct {
		name      string
		values    []int
		target    int
		wantLeft  int
		wantRight int
		wantFound bool
	}{
		{"finds the pair the lesson animates", []int{1, 2, 4, 7, 11, 15}, 15, 2, 4, true},
		{"reports no pair for an empty slice rather than guessing", []int{}, 5, 0, 0, false},
		{"reports no pair when none sums to the target", []int{1, 2, 4, 7, 11, 15}, 100, 0, 0, false},
		{"reports no pair for a single-element slice", []int{5}, 5, 0, 0, false},
		{"handles an adjacent pair", []int{3, 4}, 7, 0, 1, true},
		{"handles a pair at both endpoints", []int{1, 9, 9, 9, 10}, 11, 0, 4, true},
		{"handles negative values", []int{-8, -3, 0, 2, 5}, -3, 0, 4, true},
		{"does not pair the same element with itself", []int{4, 8}, 8, 0, 0, false},
	}

	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			left, right, found := FindPairWithSum(c.values, c.target)
			if left != c.wantLeft || right != c.wantRight || found != c.wantFound {
				t.Errorf(
					"FindPairWithSum(%v, %d) = (%d, %d, %v), want (%d, %d, %v)",
					c.values, c.target, left, right, found, c.wantLeft, c.wantRight, c.wantFound,
				)
			}
		})
	}
}

File: lessons/dsa/two-pointers/examples/FindPairWithSumTest.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

/**
 * A plain main-method test harness: no JUnit dependency, exits nonzero and prints every failure
 * (not just the first) so a broken change is diagnosed in one run, matching this repository's
 * Python and TypeScript test style.
 */
public final class FindPairWithSumTest {

  private FindPairWithSumTest() {}

  public static void main(String[] args) {
    List<String> failures = new ArrayList<>();

    checkEquals(
        failures,
        "finds the pair the lesson animates",
        Optional.of(new int[] {2, 4}),
        FindPairWithSum.findPairWithSum(new int[] {1, 2, 4, 7, 11, 15}, 15));

    checkEquals(
        failures,
        "reports no pair for an empty array rather than guessing",
        Optional.empty(),
        FindPairWithSum.findPairWithSum(new int[] {}, 5));
    checkEquals(
        failures,
        "reports no pair when none sums to the target",
        Optional.empty(),
        FindPairWithSum.findPairWithSum(new int[] {1, 2, 4, 7, 11, 15}, 100));
    checkEquals(
        failures,
        "reports no pair for a single-element array",
        Optional.empty(),
        FindPairWithSum.findPairWithSum(new int[] {5}, 5));

    checkEquals(
        failures,
        "handles an adjacent pair",
        Optional.of(new int[] {0, 1}),
        FindPairWithSum.findPairWithSum(new int[] {3, 4}, 7));
    checkEquals(
        failures,
        "handles a pair at both endpoints",
        Optional.of(new int[] {0, 4}),
        FindPairWithSum.findPairWithSum(new int[] {1, 9, 9, 9, 10}, 11));
    checkEquals(
        failures,
        "handles negative values",
        Optional.of(new int[] {0, 4}),
        FindPairWithSum.findPairWithSum(new int[] {-8, -3, 0, 2, 5}, -3));
    checkEquals(
        failures,
        "does not pair the same element with itself",
        Optional.empty(),
        FindPairWithSum.findPairWithSum(new int[] {4, 8}, 8));

    if (!failures.isEmpty()) {
      for (String failure : failures) {
        System.err.println("FAIL: " + failure);
      }
      System.exit(1);
    }

    System.out.println("All checks passed.");
  }

  private static void checkEquals(
      List<String> failures, String name, Optional<int[]> expected, Optional<int[]> actual) {
    boolean equal =
        expected.isPresent() == actual.isPresent()
            && (expected.isEmpty() || Arrays.equals(expected.get(), actual.get()));
    if (!equal) {
      failures.add(name + ": expected " + describe(expected) + " but got " + describe(actual));
    }
  }

  private static String describe(Optional<int[]> value) {
    return value.map(Arrays::toString).orElse("empty");
  }
}
